nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/services/azure/storage/ADLSCredentialsControllerService.java

```diff
@@ -97,16 +142,38 @@
         boolean sasTokenSet = StringUtils.isNotBlank(validationContext.getProperty(AzureStorageUtils.PROP_SAS_TOKEN).getValue());
         boolean useManagedIdentitySet = validationContext.getProperty(USE_MANAGED_IDENTITY).asBoolean();
 
-        if (!onlyOneSet(accountKeySet, sasTokenSet, useManagedIdentitySet)) {
-            StringJoiner options = new StringJoiner(", ")
-                .add(AzureStorageUtils.ACCOUNT_KEY.getDisplayName())
-                .add(AzureStorageUtils.PROP_SAS_TOKEN.getDisplayName())
-                .add(USE_MANAGED_IDENTITY.getDisplayName());
+        boolean servicePrincipalTenantIdSet = StringUtils.isNotBlank(validationContext.getProperty(SERVICE_PRINCIPAL_TENANT_ID).getValue());
+        boolean servicePrincipalClientIdSet = StringUtils.isNotBlank(validationContext.getProperty(SERVICE_PRINCIPAL_CLIENT_ID).getValue());
+        boolean servicePrincipalClientSecretSet = StringUtils.isNotBlank(validationContext.getProperty(SERVICE_PRINCIPAL_CLIENT_SECRET).getValue());
+        boolean servicePrincipalClientCertificateSet = validationContext.getProperty(SERVICE_PRINCIPAL_CLIENT_CERTIFICATE).isSet();
+
+        boolean servicePrincipalSet = servicePrincipalTenantIdSet || servicePrincipalClientIdSet || servicePrincipalClientSecretSet || servicePrincipalClientCertificateSet;
```

Comment:
Instead of using the presence of any of these properties to imply Service Principal Authentication, what about introducing one more property named something like `Authentication Type`?  The value of that property could take one of an enumerated list of values.  With that property in place, the remaining Service Principal properties could use the `dependsOn` feature of Property Descriptors.  This would provide a cleaner user experience and should also make the determination of authentication type easier to follow.

Reply:
Using `Authentication Type` with `dependsOn` would be great here. I just don't know how we could initialize this new field in existing flows. We cannot use a default value there. If `Account Key` property is filled in, then the `Authentication Type` should be initialized as "Account Key", if `SAS Token` is filled in, then "SAS Token", etc.
`Managed Identity` is more problematic because if `Authentication Type` is "Managed Identity", then no more property needed. The existing `Use Managed Identity` property with true/false would be redundant / inconsistent but property deletion is not really possible.

Reply:
That makes sense, what about having a default value of `AUTO` that would preserve existing behavior, and having other values that would make the Authentication Type explicit?

Reply:
I like this, but for an existing flow the UX will be a bit weird at first because all the fields will initially be hidden I think.

I was trying to think through whether there's a way to repurpose "Use Managed Identity" for this without being excessively dirty? I think there is. It's possible to add allowable values without breaking anything, display name can of course change, if you change to `AllowableValue` the existing values will look natural in the UI, and everything except the `name()` value in the code might look pretty much like this is how it originally was done.

Can we get past having this property still being identified as "storage-use-managed-identity"?

Reply:
"_the UX will be a bit weird at first because all the fields will initially be hidden_" : I don't think so because the default AUTO would mean `all properties displayed`
if I understand @exceptionfactory 's idea correctly 

Regarding repurposing "Use Managed Identity": we could rename the property on the UI and also the existing `true` to `Managed Identity` but `false` covers `Account Key` and `SAS Token` and we don;t know which one, so it is not clear for me how to handle it

Reply:
As @turcsanyip described, the default value of a new `Authentication Type` property would be `AUTO`, indicating that currently visible properties would be displayed and the existing logic to infer the desired authentication type would be followed.  Selecting a more specific `Authentication Type` value would hide properties that are not applicable using the depend on feature.

Reply:
That makes sense. I think it'll be easier to see how it feels with it in action.

We don't have an answer re: the redundancy with the value of Auth Type = Managed Identity and Use Managed Identity = t/f, right? Is the idea for them to co-exist and then validation ensures it's consistent?

Reply:
My idea would be:
- in case of `Auth Type = AUTO`: all props seen, the current validation running (eg. `Account Key` cannot be filled in and `Use Managed Identity` cannot be true at the same time)
- in case of `Auth Type = Account Key`: only `Account Key` prop seen, only this property validated (must be filled in), the other invisible properties can have any values (it would be weird to show warnings for those properties in this case)
- in case of `Auth Type = Managed Identity`: no other property can be seen, neither `Use Managed Identity`, no further validation needed, this Auth Type simply turns on Managed Identity authentication regardless of the value of the invisible `Use Managed Identity` property

With these rules, the visible properties are always consistent. The invisible ones may not be but that seems to be acceptable for me.

What do you think of it?

Reply:
@turcsanyip That sounds like a good approach.